### PR TITLE
Allow Public functions in interfaces

### LINKED ISSUE #13890: Support public functions in interfaces
## Abstract

Currently, interface only allow functions declared as external. In the implementation, the function can be declared public to extend the scope. However sometimes complexify inheritance patterns (see Motivation bellow). By allowing functions to be declared as public in interfaces, we would improve the composability of contracts

## Motivation

Lets consider the following code

```solidity
interface ISomething {
    function doSomething(bytes32) external;
}

contract Something is ISomething {
    event gotSomething(address indexed sender, bytes32 value);

    function doSomething(bytes32 value) public override {
        emit gotSomething(msg.sender, value);
    }
}

contract SomethingElse is ISomething {
    bytes32[] public values;

    function doSomething(bytes32 value) public override {
        values.push(value);
    }
}

abstract contract SomethingExtra is ISomething {
    function doSomething(string calldata value) external {
        doSomething(keccak256(bytes(value)));
    }
}

contract Composed1 is Something, SomethingExtra {}
contract Composed2 is SomethingElse, SomethingExtra {}
```

This doesn't compile because `SomethingExtra.doSomething(string)` can't access `ISomething.doSomething(bytes32)` that is declared external. A simple fix here would be to do 

```diff
-        doSomething(keccak256(bytes(value)));
+        this.doSomething(keccak256(bytes(value)));
```

but when using Composed1, the "sender" of events triggered using the string variant would not be correct. Also, calling `this.doSomething` is possibly more expensive because of redirection gas cost when working with a proxy.

Another option is to do

```diff
-interface ISomething {
-    function doSomething(bytes32) external;
+abstract contract ISomething {
+    function doSomething(bytes32) public virtual;
```

and this would be seamless here, but from my experience, this often result in override specification nightmare, because the "interface" everything inherit from is actually not declared as an interface.

## Specification

Allow functions in interfaces to be declared as public. There function will have to be implemented publicly. Implementing them using an external function would be an error. Any contract that inherit the interface would be able to call that function. In particular, abstract contract that inherit the interface but don't implement the function would still be able to call it.

## Backwards Compatibility

This doesn't affect the code that currently compile as the behaviour of the compiler with external functions declared in interfaces.

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #13890

- Should "default" visibility be accepted? Is it equivalent to public or do we want public to be explicit?
- I'm not sure why, but some test fail ... can you help me identify why ?

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -547,8 +547,8 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 
 		if (_function.isConstructor())
 			m_errorReporter.typeError(6482_error, _function.location(), "Constructor cannot be defined in interfaces.");
-		else if (_function.visibility() != Visibility::External)
-			m_errorReporter.typeError(1560_error, _function.location(), "Functions in interfaces must be declared external.");
+		else if (_function.visibility() != Visibility::External || _function.visibility() != Visibility::Public)
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: > I'm not sure why, but some test fail ... can you help me identify why ?

No visibility will match here. This should be:

```suggestion
		else if (_function.visibility() != Visibility::External && _function.visibility() != Visibility::Public)
```

**Review by cameel — COMMENTED** [review]
One thing. Not sure we'll actually accept this PR. What I said in https://github.com/ethereum/solidity/issues/13890#issuecomment-1405218045 was just my opinion and I started a discussion internally in the team to get a decision but looks like not everyone agrees that this is an acceptable feature. This will have to wait until we actually get some consensus in the team.

Generally, any issue that does not have a "desirability" label (`must have`, `must have eventually`, `should have`, `nice to have`) is undecided. Perhaps we should introduce an explicit `needs decision` label to make it clear. We did not do it mainly because it's the default state and we wanted to avoid having to put it on almost everything.

**Comment by Amxx** [comment]
> One thing. Not sure we'll actually accept this PR. What I said in [#13890 (comment)](https://github.com/ethereum/solidity/issues/13890#issuecomment-1405218045) was just my opinion and I started a discussion internally in the team to get a decision but looks like not everyone agrees that this is an acceptable feature. This will have to wait until we actually get some consensus in the team.

I fully understand. 

The  `good first issue candidate`, `low effort` and `low impact` labels made me think this was a good opportunity for me give it a try, and see if I can contribute to the compiler. If you end up doing it (which I hope) then maybe you'll appreciate having the PR ready. Otherwise no worry, I'd still be happy I gave it a try.

**Comment by cameel** [comment]
Sure, no worries. Thanks for taking the time anyway.

We actually introduced `good first issue candidate` at some point specifically because we had some issues that looked like it would be a `good first issue` eventually but were blocked for various reasons and had to be held back :) I did intend to make this one into that once it's accepted for implementation.

We're planning to restart the design calls in the coming weeks and do more decisions on pending issues. There was a lot of chaos in the last months so we have a bit of a backlog now.

**Comment by calmacfadden** [comment]
While the changes make sense, it creates additional questions on whether the definition in the interface should be enforced in the contract that implements the interface. 

Currently when an interface defined an external function it can be implemented as external or public. If the interface allows for both external and public, should the contract enforce this?

**Comment by Amxx** [comment]
> Currently when an interface defined an external function it can be implemented as external or public. If the interface allows for both external and public, should the contract enforce this?

IMO:
- functions declared as external in the interface can be implemented as either external or public
- functions declared as public in the interface can only be implemented as public.

This is already the way overriding works between (abstract) contracts:

```solidity
abstract contract AbstractPublic {
    function get() public virtual;
}

abstract contract AbstractExternal {
    function get() external virtual;
}

contract PublicPublic is AbstractPublic {
    function get() public override {} // ok
}

contract ExternalExternal is AbstractExternal {
    function get() external override {} // ok
}

contract ExternalPublic is AbstractExternal {
    function get() public override {} // ok
}
contract PublicExternal is AbstractPublic {
    function get() external override {} // TypeError: Overriding function visibility differs.
}
```

**Comment by cameel** [comment]
@NunoFilipeSantos This one is really blocked on a design decision rather than missing a review. @ekpyron was generally against it (there are some things we could potentially do as a part of #13727) but wanted to hear more opinions before we go for closing it: https://github.com/ethereum/solidity/issues/13890#issuecomment-1408642168

**Comment by ekpyron** [comment]
We concluded that we won't merge this PR as is, but we acknowledge the underlying issue and hope to be able to provide better mechanisms around that with https://github.com/ethereum/solidity/issues/13727, even though this will require a bit of more patience, for which I apologize.

**Pull request closed without merging**